src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemName.cs

```diff
@@ -242,6 +264,19 @@ internal static bool MatchPattern(string expression, ReadOnlySpan<char> name, bo
                             }
                             else
                             {
+                                if (expressionChar == '\\')
+                                {
+                                    // Escape character, try to move the expression forward again and match literally.
+                                    if (++expressionOffset == expression.Length)
+                                    {
+                                        currentMatches[currentMatch++] = maxState;
+                                        goto ExpressionFinished;
+                                    }
+
+                                    currentState = expressionOffset * 2 + 2;
```

Comment:
Can you explain the `* 2` ? I see this is already in the code but I'm having trouble picturing why the offset is getting multiplied and divided.

Reply:
It uses even/odd positioning to track multiple match states. It melts my head as well, but this is a copy of the actual Windows algorithm so I've kept it logically identical. (Tweaked for C# and our use- I added escaping support and the ability to skip the extended wildcards.)

Reply:
Needs more comments

Reply:
I don't know that I can write an explanation that wouldn't be more confusing. This came from http://source.dot.net/#System.IO.FileSystem.Watcher/System/IO/PatternMatcher.cs,785116727bd6631b. I did what I could to put better names on things and I couldn't see how the long form comments at the top were more helpful than confusing. I'll give it more thought. Perhaps someone with better algorithm-fu can assist.